### PR TITLE
optimize-selectors qSA before Sizzle

### PR DESCRIPTION
**Pull request by mr21**
Hi :)

Here a first **draft** (who probably contain few mistakes) for the issue #666.

### CODE REVIEW

**Comment by agcolom** [comment]
@dmethvin @timmywil It would be truly wonderful if you could review this PR. Thank you :-)

File: page/performance/optimize-selectors.md [hunk 1]
```diff
@@ -7,19 +7,28 @@
 
 Selector optimization is less important than it used to be, as more browsers implement `document.querySelectorAll()` and the burden of selection shifts from jQuery to the browser. However, there are still some tips to keep in mind.
 
-## ID-Based Selectors
+## Save calls to `querySelectorAll`
 
-Beginning your selector with an ID is always best.
+```
```
Comment (AurelioDeRosa) [inline page/performance/optimize-selectors.md]: I think this  example might need a small introduction.  Even 1-2 sentences should do it.

File: page/performance/optimize-selectors.md [hunk 2]
```diff
@@ -7,19 +7,28 @@
...
+$( ".container div .robotarm" );
 ```
-// Fast:
-$( "#container div.robotarm" );
 
-// Super-fast:
-$( "#container" ).find( "div.robotarm" );
+## Try to segregate the nonstandard parts of your selection
+
+When you select elements, jQuery will call `querySelectorAll` with your selection. If `querySelectorAll` throws an error, jQuery will refer to its Sizzle engine. So, if you are using at least one of these nonstandard pseudo-classes: `:contains()`, `:has`, `:even`, `:submit`, etc. You will not take advantage of the native `querySelectorAll`.
```
Comment (AurelioDeRosa) [inline page/performance/optimize-selectors.md]: I think this sentence

> if you are using at least one of these nonstandard pseudo-classes: `:contains()`, `:has`,

can be tweaked a bit

> if you are using at least one of the nonstandard pseudo-classes such as `:contains()`, `:has`,

File: page/performance/optimize-selectors.md [hunk 3]
```diff
@@ -7,19 +7,28 @@
...
+
 ```
+// A long selection with nonstandard pseudo-classes inside:
```
Comment (AurelioDeRosa) [inline page/performance/optimize-selectors.md]: Maybe we can remove the colon at the end of this comment and the next for consistency. This isn't very important, though.

**Comment by AurelioDeRosa** [comment]
Hi @Mr21. Are you still interested in this PR?

**Comment by Mr21** [comment]
... I'm really sorry, yes I'm still interested.
I'll push ASAP your change

**Comment by kswedberg** [comment]
@Mr21 Make sure you rebase from here before continuing your work. I merged a PR based on another issue and removed/rearranged a bunch of stuff, so you might have conflicts.

**Comment by Mr21** [comment]
Understood :)

**Comment by Mr21** [comment]
Hi :)

So, I have rebased and repush again my commit with the @AurelioDeRosa's modifications.
Except that I didn't add yet a small introduction in place of this removed sentence:
*Beginning your selector with an ID is always best.*

**Comment by AurelioDeRosa** [comment]
Hi @Mr21.

With "yet" you mean that you'll add another commit to update the PR?

**Comment by Mr21** [comment]
Hi :)

For the moment I have replaced this:

----

#### ID-Based Selectors
Beginning your selector with an ID is a safe bet.
``` javascript
// Fast:
$( "#container div.robotarm" );

// Super-fast:
$( "#container" ).find( "div.robotarm" );
```
With the first approach, jQuery queries the DOM using `document.querySelectorAll()`. With the second, jQuery uses `document.getElementById()`, which is faster, although the speed improvement may be diminished by the subsequent call to `.find()`.

---
by this:

#### Save calls to `querySelectorAll`
``` javascript
// If in your HTML there are 2 .container with 5 div in each,
// this line will call querySelectorAll 13 times (1 + 2 + 2*5).
$( ".container" ).children( "div" ).find( ".robotarm" );
  
// Against only 1 call with this:
$( ".container div .robotarm" );
```
---

But I don't know what I have to write to introduce my example.
If you have an idea, I will commit it quickly.

Thanks for your patience :)

**Comment by Krinkle** [comment]
I've confirmed from the previous CLA data that this PR was indeed signed as the bot originally indicated:

https://github.com/jquery/learn.jquery.com/commit/cb2c81b5816f112368c34090f4214bf613c1b881
```
/var/www/jquery-license/output/jquery/learn.jquery.com/cb/cb2c81b5816f112368c34090f4214bf613c1b881.json
```
```json
{
  "data":{
    "commits":[{
      "hash":"cb2c81b5816f112368c34090f4214bf613c1b881",
      "email":"thomastortorini@gmail.com",
      "name":"Thomas Tortorini"
    }],
    "neglectedCommits":[],
    "neglectedAuthors":[]
  },
  "state":"success",
  "description":"All authors have signed the CLA"
}
```

**Pull request merged**